[PyTorch][JIT] Put `pc` in a register during the interpreter loop

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54196 [PyTorch][JIT] Put `pc` in a register during the interpreter loop**
* #54195 [PyTorch][JIT] RFC: Store function return_pc & unify current pc
* #54110 [PyTorch][JIT] Less shared_ptr use in dictConstruct
* #54124 [PyTorch] Avoid extra intrusive_ptr copy in IValue::toIntrusivePtr

It's much better if we don't have to do a memory access to
increment the `pc` every iteration. The cost is some code complexity
and being careful to save/restore the `pc` when we `WAIT`.

Differential Revision: [D27124849](https://our.internmc.facebook.com/intern/diff/D27124849/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27124849/)!

swolchok committed:
[PyTorch][JIT] Put `pc` in a register during the interpreter loop

It's much better if we don't have to do a memory access to
increment the `pc` every iteration. The cost is some code complexity
and being careful to save/restore the `pc` when we `WAIT`.

Differential Revision: [D27124849](https://our.internmc.facebook.com/intern/diff/D27124849/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27124849/)!

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/runtime/interpreter.cpp:
```diff
@@ -1658,7 +1673,7 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
           // undo that to report the call node
           ? frames[i + 1].return_pc - 1
-          : this->pc;
+          : this->saved_pc_;
```
`formatStackTrace` is being called before saved_pc_ is restored to the current pc value in the exception handling code. So I think as currently written this is going to misreport the location of all exceptions. Generally I am worried about how fragile this approach will be over time. For context, the interpreter used to have an `ActivateFrame` object lived only in the interpreter loop function and held the PC as well as some pre-computed values. We removed it because it was more difficult to maintain than any performance it provided.

zdevito reviewed (commented):
I am OK will making the interpreter more complicated to improve performance, but I feel like there needs to be a pretty strong justification that this particular change is noticeable, given the potential for bugs (see below for one possible instance). There is a lot of other overhead in the loop (atomic increments), so I am not sure a priori this change will be noticeable.

swolchok commented:
> I feel like there needs to be a pretty strong justification that this particular change is noticeable

What measurements do you want to see? For example, check out the Facebook-internal diff -- it has a 1% *overall* win (and probably 20-25% of time in runImpl) on an arbitrarily-selected workload.

swolchok committed:
fix a couple bugs on "[PyTorch][JIT] Put `pc` in a register during the interpreter loop"


It's much better if we don't have to do a memory access to
increment the `pc` every iteration. The cost is some code complexity
and being careful to save/restore the `pc` when we `WAIT`.

Differential Revision: [D27124849](https://our.internmc.facebook.com/intern/diff/D27124849/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27124849/)!

[ghstack-poisoned]

swolchok commented:
>  it has a 1% overall win (and probably 20-25% of time in runImpl) on an arbitrarily-selected workload.

Correction: those numbers should be taken with a very hefty grain of salt given the severity of at least one of the bugs I just fixed. Will up date after I finish re-measuring, and will abandon if it turns out this doesn't do much.
